default/python/AI/PolicyAI.py

```diff
@@ -195,6 +286,59 @@ def _calculate_empire_output(self) -> _EmpireOutput:
         debug(f"Empire output: {result}")
         return result
 
+    def _rate_diversity(self) -> float:
+        """Rate diversity."""
+        if diversity not in self._empire.availablePolicies:
+            return 0.0
+        diversity_value = len(get_empire_planets_by_species()) - fo.getNamedValue("PLC_DIVERSITY_THRESHOLD")
+        # diversity affects stability, but also gives a bonus to research-focused planets and a little influence,
+        diversity_scaling = fo.getNamedValue("PLC_DIVERSITY_SCALING")
+        research_priority = self._aistate.get_priority(PriorityType.RESOURCE_RESEARCH)
+        rating = (
+            self._rate_opinion(diversity)
+            # 4 for global influence
+            + diversity_scaling * diversity_value * (self._num_populated + 4 + max(0, research_priority - 20))
+        )
+        debug(
+            f"_rate_diversity: rating={rating}. diversity_value={diversity_value}, "
+            f"research priority={research_priority}"
+        )
+        return rating
+
+    def _rate_artisans(self) -> float:
```

Comment:
First I thought: why not...
```
for pid in planets:
    rating += self._rate_artisan_planet(...)
```
...but to do that I'll have to pass a whole screen line of arguments to that function.
I'd rather improve the documentation. I'll improved the function a bit, too.

Reply:
Well, not if I put the getNamedValue calls into the function. The values should one day become global variables anyway.

Reply:
> but to do that I'll have to pass a whole screen line of arguments to that function

This is a sign of complexity.  When you spread it across the big function it's not so visible, but when you start extraction, you will hit it hard.

That is exactly what I want to avoid in a new code.